### PR TITLE
token-client enhancement with X-zid header

### PR DESCRIPTION
**Pull request by liga-oz**

### CODE REVIEW

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java [hunk 1]
```diff
@@ -187,6 +187,31 @@ public OAuth2TokenResponse retrieveAccessTokenViaJwtBearerTokenGrant(URI tokenEn
...
+		assertNotNull(xZidHeader, "X-zid header is required");
+
+		Map<String, String> parameters = new RequestParameterBuilder()
+				.withGrantType(GRANT_TYPE_JWT_BEARER)
+				.withClientCredentials(clientCredentials)
+				.withToken(token)
+				.withOptionalParameters(optionalParameters)
+				.buildAsMap();
+
+		HttpHeaders headers = HttpHeadersFactory.createWithXzidHeader(xZidHeader);
```
Comment (nenaraab) [inline token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java:207]: i would suggest sth like:
HttpHeaders headers = HttpHeadersFactory.createWithoutAuthorizationHeader().with(HttpHeaders.X_ZID, xZidHeader);

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/client/OAuth2TokenService.java [hunk 2]
```diff
@@ -210,4 +210,29 @@ default OAuth2TokenResponse retrieveAccessTokenViaJwtBearerTokenGrant(URI tokenE
...
+	 *            token.
+	 * @param token
+	 *            the JWT token identifying representing the user to be
+	 *            authenticated
+	 * @param optionalParameters
+	 *            optional request parameters, can be null.
+	 * @param disableCache
+	 *            setting to true disables the token cache for this request.
+	 * @param xZid
+	 *            zone id of IAS tenant
```
Comment (nenaraab) [inline token-client/src/main/java/com/sap/cloud/security/xsuaa/client/OAuth2TokenService.java:227]: i think the IAS is not necessary...

File: spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java [hunk 3]
```diff
@@ -55,11 +56,11 @@ public boolean isXsuaaToken(String encodedJwtToken) {
 	 * Request a Xsuaa token using Ias token as a grant.
 	 *
 	 * @param iasToken
-	 *            encoded IAS token
+	 *            decoded IAS token
```
Comment (nenaraab) [inline spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java:59]: i would remove "decoded"

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java [hunk 4]
```diff
@@ -187,6 +187,31 @@ public OAuth2TokenResponse retrieveAccessTokenViaJwtBearerTokenGrant(URI tokenEn
 		return getOAuth2TokenResponse(tokenEndpoint, headers, parameters, subdomain, disableCacheForRequest);
 	}
 
+	public OAuth2TokenResponse retrieveAccessTokenViaJwtBearerTokenGrant(URI tokenEndpoint,
```
Comment (nenaraab) [inline token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java:190]: I miss the @override annotation

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java (hunk 1 above)
Comment (nenaraab) [inline token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java:207]: I would prefer  HttpHeadersFactory.createWithoutAuthorizationHeader().withHeader(X_ZID, xZidHeader);

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/http/HttpHeadersFactory.java [hunk 5]
```diff
@@ -34,6 +34,10 @@ public static HttpHeaders createWithoutAuthorizationHeader() {
 		return createFromHeaders(createDefaultHeaders());
 	}
 
+	public static HttpHeaders createWithXzidHeader(String xZidValue) {
```
Comment (nenaraab) [inline token-client/src/main/java/com/sap/cloud/security/xsuaa/http/HttpHeadersFactory.java:37]: see feedback above, would rename parameter to "zoneId"

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/tokenflows/UserTokenFlow.java [hunk 6]
```diff
@@ -57,13 +59,27 @@
...
 	 * Sets the JWT token that should be exchanged for another JWT token.
 	 *
-	 * @param token
+	 * @param encodedToken
 	 *            - the JWT token.
 	 * @return this builder object.
 	 */
-	public UserTokenFlow token(String token) {
+	public UserTokenFlow token(String encodedToken) {
+		assertNotNull(encodedToken, "Token must not be null.");
```
Comment (nenaraab) [inline token-client/src/main/java/com/sap/cloud/security/xsuaa/tokenflows/UserTokenFlow.java:67]: Token must not be null. -> Encoded Token must not be null.

File: token-client/src/test/java/com/sap/cloud/security/xsuaa/http/HttpHeadersFactoryTest.java [hunk 7]
```diff
@@ -29,4 +30,11 @@ public void createWithAuthorizationBearerHeader_containsDefaultHeadersAndToken()
 		assertThat(headers.getHeaders()).contains(tokenHeader);
 	}
 
+	@Test
+	public void createWithXzidHeader() {
```
Comment (nenaraab) [inline token-client/src/test/java/com/sap/cloud/security/xsuaa/http/HttpHeadersFactoryTest.java:34]: This will then change...

**Review by nenaraab — CHANGES_REQUESTED** [review]
Hi @liga-oz 

wow! Very good and looks quite clean!!!
I have some minor improvement requests, mainly in context of the HttpHeader creation...
Thanks a lot!

Btw.: Have you planned to use the TokenFactory in a separate PR?

Nena

File: token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java (hunk 1 above)
Comment (liga-oz) [reply token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java:207]: to chain the methods I would need to convert the methods to instance methods, or is there another way for method chaining in static context:question:

Comment (nenaraab) [reply token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java:207]: Hi @liga-oz 
The "with" method will become an instance method of HttpHeaders, which then returns itself as instance of type HttpHeaders)...

Comment (liga-oz) [reply token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java:207]: ah, got it thanks 👍

**Review by nenaraab — APPROVED** [review]

**Pull request merged**